### PR TITLE
Plfm 4247 e

### PR DESCRIPTION
**Pull request by john-hill**

### CODE REVIEW

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLUtils.java [hunk 1]
```diff
@@ -1714,4 +1716,57 @@ public static void writeAnnotationDtoToPreparedStatement(PreparedStatement ps, A
...
+		boolean useBindVariables = false;
+		buildHeaders(builder, headers, useBindVariables);
+		builder.append(" VALUES ");
+		useBindVariables = true;
+		buildHeaders(builder, headers, useBindVariables);
+		return builder.toString();
+	}
+	
+	static void buildHeaders(StringBuilder builder, String[] headers, boolean useBindVariables) {
+		builder.append(" (");
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/SQLUtils.java:1740]: consider stringjoiner

File: lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLUtilsTest.java [hunk 2]
```diff
@@ -2394,4 +2395,20 @@ public void testWriteAnnotationDtoToPreparedStatementNegativeInfinity() throws S
...
 	}
+	
+	@Test
+	public void testCreateInsertViewFromSnapshot() {
+		String[] headers = new String[] {"foo","bar"};
+		tableId = IdAndVersion.parse("syn999.23");
+		// call under test
+		String sql = SQLUtils.createInsertViewFromSnapshot(tableId, headers);
+		assertEquals("INSERT INTO T999_23 (foo,bar) VALUES  (?,?)", sql);
+	}
```
Comment (john-hill) [inline lib/lib-table-cluster/src/test/java/org/sagebionetworks/table/cluster/SQLUtilsTest.java:2406]: add case with only one value.

File: lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/TableIndexDAO.java [hunk 3]
```diff
@@ -430,4 +431,17 @@ public void setIndexVersionAndSchemaMD5Hex(IdAndVersion tableId, Long viewCRC,
 	 * @return
 	 */
 	public void streamSynapseStorageStats(Callback<SynapseStorageProjectStats> callback);
+
+	/**
+	 * Populate a view from a snapshot.
+	 * 
+	 * @param idAndVersion
+	 * @param input
```
Comment (john-hill) [inline lib/lib-table-cluster/src/main/java/org/sagebionetworks/table/cluster/TableIndexDAO.java:439]: document csv input

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableViewManagerImpl.java [hunk 4]
```diff
@@ -289,50 +300,97 @@ public void createOrUpdateViewIndex(IdAndVersion idAndVersion, ProgressCallback
...
+		Set<Long> allContainersInScope = tableManagerSupport.getAllContainerIdsForViewScope(idAndVersion,
+				viewTypeMask);
+		return indexManager.populateViewFromEntityReplication(idAndVersion.getId(), viewTypeMask,
+				allContainersInScope, viewSchema);
+	}
+
+	/**
+	 * Populate the view table from csv snapshot.
+	 * @param idAndVersion
+	 * @param indexManager
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableViewManagerImpl.java:367]: comment return

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableViewManagerImpl.java [hunk 5]
```diff
@@ -289,50 +300,97 @@ public void createOrUpdateViewIndex(IdAndVersion idAndVersion, ProgressCallback
...
+	 */
+	long populateViewFromSnapshot(IdAndVersion idAndVersion, TableIndexManager indexManager) {
+		ViewSnapshot snapshot = viewSnapshotDao.getSnapshot(idAndVersion);
+		File tempFile = null;
+		try {
+			tempFile = fileProvider.createTempFile("ViewSnapshotDownload", ".csv.gzip");
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableViewManagerImpl.java:373]: .csv.gzip constant

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableManagerSupportImpl.java [hunk 6]
```diff
@@ -331,13 +335,18 @@ public static ObjectType getObjectTypeForEntityType(EntityType type) {
 	 */
 	@Override
 	public Long calculateViewCRC32(IdAndVersion idAndVersion) {
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableManagerSupportImpl.java:337]: return little long

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/table/TableManagerSupportTest.java [hunk 7]
```diff
@@ -621,6 +624,18 @@ public void calculateFileViewCRC32(){
...
+	
+	@Test
+	public void calculateFileViewCRC32Veserion(){
+		idAndVersion = IdAndVersion.parse("syn123.45");
+		Long snapshotId = 33L;
+		when(mockViewSnapshotDao.getSnapshotId(idAndVersion)).thenReturn(snapshotId);
+		// call under test
+		Long crcResult = manager.calculateViewCRC32(idAndVersion);
+		assertEquals(snapshotId, crcResult);
+		verify(mockReplicationMessageManager, never()).pushContainerIdsToReconciliationQueue(any());
```
Comment (john-hill) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/table/TableManagerSupportTest.java:638]: verifyGetSnapshot

**Review by marcomarasca — APPROVED** [review]

**Pull request merged**